src/lib/zf/tx_warm.c

```diff
@@ -42,15 +36,9 @@ void disable_tx_warm(struct zf_tx* tx, zf_tx_warm_state* state)
   ef_vi* vi = zf_stack_nic_tx_vi(st_nic);
   zf_assert_flags(st->flags, ZF_STACK_FLAG_TRANSMIT_WARM_ENABLED);
   zf_log_stack_trace(st, "%s: TX warm disabled\n", __func__);
-  /* TODO better way to distuinguish X2 vs X3 style CTPIO */
-  if( vi->nic_type.arch == EF_VI_ARCH_EFCT ) {
-    efct_vi_stop_transmit_warm(vi);
-  }
-  else {
-    ef_vi_stop_transmit_warm(vi, &state->ef_vi_state[tx->path.nicno]);
-    if( state->ctpio_warm_buf_id != PKT_INVALID )
-      zf_pool_free_pkt(&st->pool, state->ctpio_warm_buf_id);
-  }
+  ef_vi_stop_transmit_warm(vi, &state->ef_vi_state[tx->path.nicno]);
+  if( state->ctpio_warm_buf_id != PKT_INVALID )
+    zf_pool_free_pkt(&st->pool, state->ctpio_warm_buf_id);
```

Comment:
Is it possible for this to be called on efct? If so, will this do anything? Before this patch this was not called and I'm wondering if there was a reason for that.

Reply:
It looks safe, `state->ctpio_warm_buf_id = PKT_INVALID;` unless `vi->vi_flags & EF_VI_TX_CTPIO && vi->nic_type.arch != EF_VI_ARCH_EFCT`, in which case it's assigned as a result of `zft_alloc_pkt`.

Reply:
The code in the enable function is writing to avoid calling zf_pool_free_pkt on efct. If you called free on a pkt not assigned that would cause issues but on efct the buf_id will be PKT_INVALID.  The ef_vi_state is ignored on EFCT so makes no difference.